packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts

```diff
@@ -0,0 +1,184 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { type IMemoryTestObject, benchmarkMemory } from "@fluid-tools/benchmark";
+
+import { SharedMatrix } from "../../index.js";
+import { createLocalMatrix } from "../utils.js";
+
+describe("SharedMatrix memory usage", () => {
+	const matrixValue = "cellValue";
+	// The test matrix's size will be 10*10, 100*100, 1000*1000.
+	const matrixSizes = [10, 100, 1000];
+
+	// The number of operations to perform on the matrix.
+	const operationCounts = [10, 100, 1000];
+
+	beforeEach(async () => {});
+
+	afterEach(() => {});
+
+	for (const matrixSize of matrixSizes) {
+		describe(`Size of ${matrixSize}*${matrixSize} SharedMatrix`, () => {
+			for (const count of operationCounts) {
+				// Test the memory usage of the SharedMatrix for inserting a column in the middle for a given number of times.
+				benchmarkMemory(
+					new (class implements IMemoryTestObject {
+						readonly title = `Insert a column in the middle ${count} times`;
+						private localMatrix: SharedMatrix | undefined;
+
+						async run(): Promise<void> {
+							for (let i = 0; i < count; i++) {
+								this.localMatrix?.insertCols(Math.floor(matrixSize / 2), 1);
+							}
+						}
+
+						beforeIteration(): void {
+							this.localMatrix = createLocalMatrix({
+								id: "testLocalMatrix",
+								size: matrixSize,
+								initialValue: matrixValue,
+							});
+						}
+					})(),
+				);
+
+				// Test the memory usage of the SharedMatrix for removing a column for a given number of times.
+				benchmarkMemory(
+					new (class implements IMemoryTestObject {
+						readonly title = `Remove the first column ${count} times`;
+						private localMatrix: SharedMatrix | undefined;
+
+						async run(): Promise<void> {
+							for (let i = 0; i < count; i++) {
+								this.localMatrix?.removeCols(0, 1);
```

Comment:
We should be careful not to use the `?.` operator when performing operations on the matrix. If initialization breaks for some reason, it will result in these tests noop-ing rather than failing. I would recommend replacing any such `?.`s with `!.`s. Feel free to ignore the linter rule if eslint complains (but please leave a quick comment next to any linter disables explaining why it's safe).

Reply:
Gotcha. Updated accordingly